src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -681,7 +681,7 @@ public static long mismatch(MemorySegment srcSegment, long srcFromOffset, long s
         long dstBytes = dstToOffset - dstFromOffset;
         srcImpl.checkAccess(srcFromOffset, srcBytes, true);
         dstImpl.checkAccess(dstFromOffset, dstBytes, true);
-        if (dstImpl == srcImpl) {
+        if (dstImpl.equals(srcImpl) && srcFromOffset == dstFromOffset && srcBytes == dstBytes) {
```

Comment:
I'm also ok to completely drop this check (or move it into the instance version of the mismatch method). This also needs a test.